diff -Naur orig-openvpn-2.1_rc10/configure.ac mod-openvpn-2.1_rc10/configure.ac
--- orig-openvpn-2.1_rc10/configure.ac	2008-08-10 18:35:25.000000000 +0000
+++ mod-openvpn-2.1_rc10/configure.ac	2008-09-22 20:37:09.741152353 +0000
@@ -865,6 +865,8 @@
 	openvpn.spec
 	config-win32.h
 	images/Makefile
+	tap-win32/SOURCES
+	tap-win32/i386/OemWin2k.inf
 	service-win32/Makefile
 	install-win32/Makefile
 	install-win32/settings
diff -Naur orig-openvpn-2.1_rc10/tap-win32/tapdrvr.c mod-openvpn-2.1_rc10/tap-win32/tapdrvr.c
--- orig-openvpn-2.1_rc10/tap-win32/tapdrvr.c	2008-08-10 18:35:25.000000000 +0000
+++ mod-openvpn-2.1_rc10/tap-win32/tapdrvr.c	2008-09-22 20:37:30.144483854 +0000
@@ -39,7 +39,6 @@
 // TAP_IOCTL_CONFIG_TUN ioctl.
 //======================================================
 
-#include "../../autodefs/defs.h"
 #ifndef DDKVER_MAJOR
 #error DDKVER_MAJOR must be defined as the major number of the DDK Version
 #endif
diff -Naur orig-openvpn-2.1_rc10/version.m4 mod-openvpn-2.1_rc10/version.m4
--- orig-openvpn-2.1_rc10/version.m4	2008-09-10 06:55:52.000000000 +0000
+++ mod-openvpn-2.1_rc10/version.m4	2008-09-22 20:35:28.208245859 +0000
@@ -1,6 +1,6 @@
 dnl define the OpenVPN version
 define(PRODUCT_VERSION,[2.1_rc10])
 dnl define the TAP version
-define(PRODUCT_TAP_ID,[tap0901])
+define(PRODUCT_TAP_ID,[tortap91])
 define(PRODUCT_TAP_WIN32_MIN_MAJOR,[9])
 define(PRODUCT_TAP_WIN32_MIN_MINOR,[1])
