diff -Naur orig-qt-4.4.3/src/network/ssl/qsslsocket_openssl_symbols.cpp mod-qt-4.4.3/src/network/ssl/qsslsocket_openssl_symbols.cpp
--- orig-qt-4.4.3/src/network/ssl/qsslsocket_openssl_symbols.cpp	2008-09-27 09:00:56.000000000 +0000
+++ mod-qt-4.4.3/src/network/ssl/qsslsocket_openssl_symbols.cpp	2008-12-30 09:53:24.191484064 +0000
@@ -327,14 +327,22 @@
     if (!ssleay32->load()) {
         // Cannot find ssleay32.dll
         delete ssleay32;
-        return pair;
+        ssleay32 = new QLibrary(QLatin1String("ssleay32-0.9.8"));
+        if (!ssleay32->load()) {
+            delete ssleay32;
+            return pair;
+        }
     }
 
     QLibrary *libeay32 = new QLibrary(QLatin1String("libeay32"));
     if (!libeay32->load()) {
-        delete ssleay32;
         delete libeay32;
-        return pair;
+        libeay32 = new QLibrary(QLatin1String("cryptoeay32-0.9.8"));
+        if (!libeay32->load()) {
+            delete libeay32;
+            delete ssleay32;
+            return pair;
+        }
     }
 
     pair.first = ssleay32;
