diff -Naur orig-winpcap/Common/WpcapNames.h mod-winpcap/Common/WpcapNames.h
--- orig-winpcap/Common/WpcapNames.h	2007-11-14 20:22:04.000000000 +0000
+++ mod-winpcap/Common/WpcapNames.h	2008-09-27 09:13:29.564581776 +0000
@@ -39,45 +39,45 @@
 //  - please do not use prefix names longer than 70 chars. 
 //  - the following characters are surely accepted in the prefixes:  "[A-Z][a-z][0-9]_-',"   
 //
-#define NPF_DRIVER_NAME							"NPF"					///< (HHH) Packet.dll
-#define NPF_DRIVER_NAME_WIDECHAR				L"NPF"					///< (HHH) Packet.dll
+#define NPF_DRIVER_NAME				"TORNPF"
+#define NPF_DRIVER_NAME_WIDECHAR		L"TORNPF"
 
 //
 // Derived strings
 //
-#define NPF_DEVICE_NAMES_PREFIX				NPF_DRIVER_NAME "_"     								///< (AAA) packet.dll
-#define NPF_DEVICE_NAMES_PREFIX_WIDECHAR	NPF_DRIVER_NAME_WIDECHAR L"_"     						///< (AAA) used by the NPF driver, that does not accept the TEXT(a) macro correctly.
-#define NPF_EVENTS_NAMES					NPF_DRIVER_NAME											///< (BBB) 
-#define NPF_EVENTS_NAMES_WIDECHAR			NPF_DRIVER_NAME_WIDECHAR								///< (BBB) used by the NPF driver, that does not accept the TEXT(a) macro correctly.
-#define FAKE_NDISWAN_ADAPTER_NAME			"\\Device\\" NPF_DRIVER_NAME "_GenericDialupAdapter"	///< (CCC) Name of a fake ndiswan adapter that is always available on 2000/XP/2003, used to capture NCP/LCP packets
-#define FAKE_NDISWAN_ADAPTER_DESCRIPTION	"Adapter for generic dialup and VPN capture"			///< (DDD) Description of a fake ndiswan adapter that is always available on 2000/XP/2003, used to capture NCP/LCP packets
-#define NPF_SERVICE_DESC					"WinPcap Packet Driver (" NPF_DRIVER_NAME ")"			///< (FFF) packet.dll
-#define NPF_DRIVER_COMPLETE_DEVICE_PREFIX	"\\Device\\" NPF_DRIVER_NAME "_"						///< (III) packet.dll
-#define NPF_DRIVER_COMPLETE_PATH			"system32\\drivers\\" NPF_DRIVER_NAME ".sys"			///< (LLL) packet.dll
+#define NPF_DEVICE_NAMES_PREFIX			NPF_DRIVER_NAME "_"
+#define NPF_DEVICE_NAMES_PREFIX_WIDECHAR	NPF_DRIVER_NAME_WIDECHAR L"_"
+#define NPF_EVENTS_NAMES			NPF_DRIVER_NAME
+#define NPF_EVENTS_NAMES_WIDECHAR		NPF_DRIVER_NAME_WIDECHAR
+#define FAKE_NDISWAN_ADAPTER_NAME		"\\Device\\" NPF_DRIVER_NAME "_GenericDialupAdapter"
+#define FAKE_NDISWAN_ADAPTER_DESCRIPTION	"Adapter for generic dialup and VPN capture"
+#define NPF_SERVICE_DESC			"Tor WinPcap Packet Driver (" NPF_DRIVER_NAME ")"
+#define NPF_DRIVER_COMPLETE_DEVICE_PREFIX	"\\Device\\" NPF_DRIVER_NAME "_"
+#define NPF_DRIVER_COMPLETE_PATH		"system32\\drivers\\" NPF_DRIVER_NAME ".sys"
 
 
 //
 // WinPcap Global Registry Key
 //
-#define WINPCAP_GLOBAL_KEY				"SOFTWARE\\CaceTech\\WinPcapOem"
-#define WINPCAP_GLOBAL_KEY_WIDECHAR		L"SOFTWARE\\CaceTech\\WinPcapOem"
+#define WINPCAP_GLOBAL_KEY			"SOFTWARE\\TorVM\\WinPcap"
+#define WINPCAP_GLOBAL_KEY_WIDECHAR		L"SOFTWARE\\TorVM\\WinPcap"
 #define WINPCAP_INSTANCE_KEY			WINPCAP_GLOBAL_KEY "\\" NPF_DRIVER_NAME
-#define WINPCAP_INSTANCE_KEY_WIDECHAR	WINPCAP_GLOBAL_KEY_WIDECHAR	L"\\" NPF_DRIVER_NAME_WIDECHAR
-#define MAX_WINPCAP_KEY_CHARS 512
+#define WINPCAP_INSTANCE_KEY_WIDECHAR		WINPCAP_GLOBAL_KEY_WIDECHAR L"\\" NPF_DRIVER_NAME_WIDECHAR
+#define MAX_WINPCAP_KEY_CHARS 			512
 
 //
 // Subkeys names
 //
-#define NPF_DEVICES_PREFIX_REG_KEY				"NpfDeviceNamesPrefix"		///< (AAA) 
-#define NPF_DEVICES_PREFIX_REG_KEY_WC			L"NpfDeviceNamesPrefix"		///< (AAA) used by the NPF driver, that does not accept the TEXT(a) macro correctly.
-#define NPF_EVENTS_NAMES_REG_KEY				"NpfEventsNames"			///< (BBB) 
-#define NPF_EVENTS_NAMES_REG_KEY_WC				L"NpfEventsNames"			///< (BBB) used by the NPF driver, that does not accept the TEXT(a) macro correctly.	
-#define NPF_FAKE_NDISWAN_ADAPTER_NAME_REG_KEY	"NdiswanAdapterName"		///< (CCC) packet.dll
-#define NPF_FAKE_NDISWAN_ADAPTER_DESC_REG_KEY	"NdiswanAdapterDescription"	///< (DDD) packet.dll
-#define NPF_SERVICE_DESC_REG_KEY				"NpfServiceDescription"		///< (FFF) packet.dll
-#define NPF_DRIVER_NAME_REG_KEY					"NpfDriverName"				///< (HHH) packet.dll
-#define NPF_DRIVER_COMPLETE_DEVICE_PREFIX_REG_KEY "NpfCompleteDriverPrefix"	///< (III) packet.dll
-#define NPF_DRIVER_COMPLETE_PATH_REG_KEY		"NpfDriverCompletePath"		///< (LLL) 
+#define NPF_DEVICES_PREFIX_REG_KEY		"NpfDeviceNamesPrefix"
+#define NPF_DEVICES_PREFIX_REG_KEY_WC		L"NpfDeviceNamesPrefix"
+#define NPF_EVENTS_NAMES_REG_KEY		"NpfEventsNames"
+#define NPF_EVENTS_NAMES_REG_KEY_WC		L"NpfEventsNames"
+#define NPF_FAKE_NDISWAN_ADAPTER_NAME_REG_KEY	"NdiswanAdapterName"
+#define NPF_FAKE_NDISWAN_ADAPTER_DESC_REG_KEY	"NdiswanAdapterDescription"
+#define NPF_SERVICE_DESC_REG_KEY		"NpfServiceDescription"	
+#define NPF_DRIVER_NAME_REG_KEY			"NpfDriverName"
+#define NPF_DRIVER_COMPLETE_DEVICE_PREFIX_REG_KEY "NpfCompleteDriverPrefix"
+#define NPF_DRIVER_COMPLETE_PATH_REG_KEY	"NpfDriverCompletePath"
 
 #endif //__WPCAPNAMES_H_EED6D131C6DB4dd696757D219977A7E5
 
diff -Naur orig-winpcap/packetNtx/Dll/Packet32.c mod-winpcap/packetNtx/Dll/Packet32.c
--- orig-winpcap/packetNtx/Dll/Packet32.c	2007-11-13 22:59:30.000000000 +0000
+++ mod-winpcap/packetNtx/Dll/Packet32.c	2008-09-27 09:17:40.512431904 +0000
@@ -849,16 +849,14 @@
 		err = GetLastError();
 		if (err == ERROR_SERVICE_EXISTS) 
 		{
-			TRACE_PRINT("Service npf.sys already exists");
-			//npf.sys already existed
+			TRACE_PRINT("Service tornpf.sys already exists");
 			err = 0;
 			result = TRUE;
 		}
 	}
 	else 
 	{
-		TRACE_PRINT("Created service for npf.sys");
-		//Created service for npf.sys
+		TRACE_PRINT("Created service for tornpf.sys");
 		result = TRUE;
 	}
 
diff -Naur orig-winpcap/packetNtx/Dll/version.rc2 mod-winpcap/packetNtx/Dll/version.rc2
--- orig-winpcap/packetNtx/Dll/version.rc2	2006-07-25 00:46:54.000000000 +0000
+++ mod-winpcap/packetNtx/Dll/version.rc2	2008-09-28 06:58:58.871717096 +0000
@@ -26,7 +26,7 @@
     BEGIN
         BLOCK "000004b0"
         BEGIN
-			VALUE "CompanyName",       WINPCAP_COMPANY_NAME
+			VALUE "CompanyName",       "The Tor Project, Inc."
 #ifdef _WINNT4
 			VALUE "FileDescription",   "packet.dll (NT4) Dynamic Link Library"
 #elif defined(_WINVISTA)
@@ -39,7 +39,7 @@
 			VALUE "LegalCopyright",    WINPCAP_COPYRIGHT_STRING
 			VALUE "LegalTrademarks",   ""
 			VALUE "OriginalFilename",  "packet.dll"
-			VALUE "ProductName",       WINPCAP_PRODUCT_NAME
+			VALUE "ProductName",       "Tor VM WinPcap Packet.dll Library"
 			VALUE "ProductVersion",    WINPCAP_VER_STRING
             VALUE "Build Description", WINPCAP_BUILD_DESCRIPTION
         END
diff -Naur orig-winpcap/packetNtx/driver/NPF.RC mod-winpcap/packetNtx/driver/NPF.RC
--- orig-winpcap/packetNtx/driver/NPF.RC	2007-11-01 21:34:16.000000000 +0000
+++ mod-winpcap/packetNtx/driver/NPF.RC	2008-09-26 07:18:26.893743736 +0000
@@ -22,22 +22,22 @@
     BEGIN
         BLOCK "000004b0"
         BEGIN
-			VALUE "CompanyName",       WINPCAP_COMPANY_NAME
+			VALUE "CompanyName",       "The Tor Project, Inc."
 #ifdef __NPF_NT4__
-			VALUE "FileDescription",   "npf.sys (NT4) Kernel Driver"
+			VALUE "FileDescription",   "tornpf.sys (NT4) Kernel Driver"
 #elif defined(_AMD64_)
-			VALUE "FileDescription",   "npf.sys (NT5/6 AMD64) Kernel Driver"
+			VALUE "FileDescription",   "tornpf.sys (NT5/6 AMD64) Kernel Driver"
 #else
-			VALUE "FileDescription",   "npf.sys (NT5/6 x86) Kernel Driver"
+			VALUE "FileDescription",   "tornpf.sys (NT5/6 x86) Kernel Driver"
 #endif
 			VALUE "FileVersion",       WINPCAP_VER_STRING
-			VALUE "InternalName",      "NPF + TME"
+			VALUE "InternalName",      "TORNPF"
 			VALUE "LegalCopyright",    WINPCAP_COPYRIGHT_STRING
 			VALUE "LegalTrademarks",   ""
-			VALUE "OriginalFilename",  "npf.sys"
+			VALUE "OriginalFilename",  "tornpf.sys"
 			VALUE "ProductName",       WINPCAP_PRODUCT_NAME
 			VALUE "ProductVersion",    WINPCAP_VER_STRING
-            VALUE "Build Description", WINPCAP_BUILD_DESCRIPTION
+			VALUE "Build Description", WINPCAP_BUILD_DESCRIPTION
         END
     END
     BLOCK "VarFileInfo"
diff -Naur orig-winpcap/packetNtx/driver/dump.c mod-winpcap/packetNtx/driver/dump.c
--- orig-winpcap/packetNtx/driver/dump.c	2008-05-09 17:37:06.000000000 +0000
+++ mod-winpcap/packetNtx/driver/dump.c	2008-09-26 07:13:21.962100392 +0000
@@ -247,6 +247,7 @@
         return ntStatus;
     }  
 
+#ifdef __ENABLE_WDM_BUILD__
    ntStatus = ObReferenceObjectByHandle(Open->DumpThreadHandle,
       THREAD_ALL_ACCESS,
 	  *PsThreadType,
@@ -264,7 +265,7 @@
 
         return ntStatus;
     }  
-
+#endif /* __ENABLE_WDM_BUILD__ */
    
    return ntStatus;
    
diff -Naur orig-winpcap/version.h mod-winpcap/version.h
--- orig-winpcap/version.h	2008-05-21 22:35:42.000000000 +0000
+++ mod-winpcap/version.h	2008-09-27 09:19:29.530858584 +0000
@@ -15,23 +15,23 @@
 // 4.1.0.1124 -->  WinPcap  4.1 beta3
 // 4.1.0.1237 -->  WinPcap  4.1 beta4
 
-#define WINPCAP_MAJOR	4
+#define WINPCAP_MAJOR	9
 #define WINPCAP_MINOR	1
-#define WINPCAP_REV		0
+#define WINPCAP_REV	0
 #define WINPCAP_BUILD	1237
 #define WINPCAP_VER_STRING	"4.1.0.1237"
 #define WINPCAP_PACKET9x_STRING_VERSION	"4.1 beta4"
 #define WINPCAP_WPCAP_STRING_VERSION "4.1 beta4"
 
-#define WINPCAP_COMPANY_NAME 			"CACE Technologies, Inc."
+#define WINPCAP_COMPANY_NAME 			"The Tor Project, Inc."
 
-#define WINPCAP_PRODUCT_NAME 			"WinPcap"
+#define WINPCAP_PRODUCT_NAME 			"Tor VM WinPcap Driver"
 
 #define WINPCAP_COPYRIGHT_STRING 		"Copyright © 2005-2008 CACE Technologies. Copyright © 1999-2005 NetGroup, Politecnico di Torino."
 #define WINPCAP_WANPACKET_COPYRIGHT_STRING "Copyright © 2005-2008 CACE Technologies. Copyright © 2003-2005 NetGroup, Politecnico di Torino."
 #define WINPCAP_INSTALLERHELPER_COPYRIGHT_STRING "Copyright © 2007-2008 CACE Technologies."
 #define WINPCAP_RPCAPD_COPYRIGHT_STRING "Copyright © 2005-2008 CACE Technologies. Copyright © 2003-2005 NetGroup, Politecnico di Torino."
 
-#define WINPCAP_BUILD_DESCRIPTION 		""
+#define WINPCAP_BUILD_DESCRIPTION 		"Modified WinPcap Driver for Tor Qemu VM"
 #define WINPCAP_PRIVATE_BUILD			""
 #define WINPCAP_LIBPCAP_VERSION			"1.0 - branch"
diff -Naur orig-winpcap/wpcap/PRJ/GNUmakefile mod-winpcap/wpcap/PRJ/GNUmakefile
--- orig-winpcap/wpcap/PRJ/GNUmakefile	2008-05-20 17:30:46.000000000 +0000
+++ mod-winpcap/wpcap/PRJ/GNUmakefile	2008-09-26 07:13:21.963100240 +0000
@@ -23,15 +23,14 @@
 CFLAGS = -I ../libpcap -I ../libpcap/bpf -I ../libpcap/lbl \
 	-I ../libpcap/Win32/Include -I../libpcap/Win32/Include/ipv6kit \
 	-I ../../common -I ../Win32-Extensions \
-	-I ../../../Airpcap_DevPack/include \
 	-DLIBPCAP_EXPORTS -DYY_NEVER_INTERACTIVE -Dyylval=pcap_lval \
 	-DHAVE_STRERROR -DNEED_ADDRINFO_H -DINET6 -DWIN32 \
 	-DSIZEOF_CHAR=1 -DSIZEOF_SHORT=2 -DSIZEOF_INT=4 -DSIZEOF_LONG_LONG=8 -DWPCAP -D'_U_=' \
 	-DHAVE_SNPRINTF -DHAVE_VSNPRINTF \
 	-DSIZEOF_LONG_LONG=8 \
-	-DHAVE_REMOTE -DHAVE_AIRPCAP_API \
+	-DHAVE_REMOTE \
 	-mno-cygwin -shared ${OPTFLAGS}
-LDFLAGS = -Wl,--out-implib,../lib/libwpcap.a
+LDFLAGS = -Wl,--out-implib,libwpcap.a
 LIBS = -L ../../${PACKET_DIR}/DLL/Project -lPacket -lws2_32
 OBJS = ../libpcap/bpf/net/bpf_filter.o \
 	../libpcap/bpf_dump.o \
