diff -Naur kamikaze-orig/package/openssl/Makefile kamikaze-mod/package/openssl/Makefile
--- kamikaze-orig/package/openssl/Makefile	2008-07-13 16:02:14.664043000 +0000
+++ kamikaze-mod/package/openssl/Makefile	2008-07-23 11:06:00.285160859 +0000
@@ -71,8 +71,8 @@
 OPENSSL_OPTIONS:= shared no-ec no-err no-fips no-hw no-threads zlib-dynamic \
 					no-sse2 no-perlasm --with-cryptodev
 else
-OPENSSL_OPTIONS:= shared no-ec no-err no-fips no-hw no-threads zlib-dynamic \
-					no-engines no-sse2 no-perlasm
+OPENSSL_OPTIONS:= shared no-ec no-err no-fips no-hw zlib-dynamic \
+					no-engines no-sse2 no-perlasm threads 386
 endif
 
 
